Create the accounts proof request

viquezclaudio opened the pull request:
Create the requests and handlers that are necessary to perform account proving.

Depends on #1462.

viquezclaudio committed:
Create the accounts proof request

Create the request and handler that is necesary to perform accounts
proving.

Co-authored-by: Jose Daniel Hernandez <jd@nimiq.com>

viquezclaudio commented:
@sisou Just to let you know. This is failing to compile to WASM because there is a dependency of the accounts crate into the database.
So we need to remove that dependency as part of the accounts refactor in order to be able to use this functionality in the web client.

sisou commented:
Would it be possible to structure the request message in a way that it allows reusing for querying stakers and validators? Right now it doesn't know the difference and I guess only returns the regular accounts. How can we query staker accounts and validator accounts?

viquezclaudio commented:
LGTM

sisou commented on consensus/src/messages/handlers.rs:
```diff
@@ -476,3 +476,28 @@ impl Handle<ResponseTransactionReceiptsByAddress, Arc<RwLock<Blockchain>>>
+        let proof = blockchain.get_accounts_proof(&self.addresses);
+
+        if proof.is_none() {
+            // If we could not generate a proof we response an empty result
```
```suggestion
            // If we could not generate a proof we respond with an empty result
```

sisou commented on consensus/src/messages/handlers.rs:
```diff
@@ -476,3 +476,28 @@ impl Handle<ResponseTransactionReceiptsByAddress, Arc<RwLock<Blockchain>>>
+            };
+        }
+
+        let block_hash = blockchain.head_hash();
```
Is this guaranteed to be the same hash that is used to generate the proof? No chance that the head is updated in this blockchain-read in the meantime?

jsdanielh replied:
Yup, the read lock is taken while this happens

sisou commented on consensus/src/messages/mod.rs:
```diff
@@ -248,3 +249,25 @@ pub struct ResponseTransactionReceiptsByAddress {
+    type Kind = RequestMarker;
+    const TYPE_ID: u16 = 215;
+    type Response = ResponseAccountsProof;
+    const MAX_REQUESTS: u32 = MAX_REQUEST_TRANSACTIONS_PROOF;
```
Are we sure we should reuse the TRANSACTIONS_PROOF limit for this request?

jsdanielh replied:
Value is correct but name is not

jsdanielh committed:
Use the Trie keys for requesting a `TrieProof`

- Use the Trie keys for requesting a `TrieProof`.
- Rename `RequestAccountsProof` to `RequestTrieProof`.
- Rename `ResponseAccountsProof` to `ResponseTrieProof`.

sisou commented on consensus/src/messages/mod.rs:
```diff
@@ -248,3 +251,25 @@ pub struct ResponseTransactionReceiptsByAddress {
 }
+
+#[derive(Clone, Debug, Serialize, Deserialize)]
+pub struct RequestTrieProof {
```
The handler for this message still needs to be registered in `consensus/src/consensus/mod.rs`.

jsdanielh replied:
Thanks. I already had forgotten about it
